vignettes/vcr.Rmd

```diff
@@ -1,191 +1,172 @@
+If your API uses JSON (like most modern APIs) you should use the `body_json` request matcher: that will parse the body as JSON so you'll get more informative messages if a new request doesn't match a request saved in a cassette.
 
-There's a number of features in this package that are not yet supported, but for which their parameters are found in the package. 
+Here's a little example with logging turned on, so you can see exactly what's happening:
 
-We've tried to make sure the parameters that are ignored are marked as such. Keep an eye out for package updates for changes in these parameters, and/or let us know you want it and we can move it up in the priority list.
+```{r}
+#| label: body-record
 
-## Examples
+get_data <- function(b = 1) {
+  req <- request(httpbin$url("/post"))
+  req <- req_body_json(req, list(x = 1, y = list(a = 1, b = b)))
+  resp <- req_perform(req)
 
-It's also possible to use vcr in an example. It's not possible to use `local_cassette()` (or `use_cassette()` inside an example), so the basic idea is to call `insert_example_cassette()` at the start of the example and `eject_cassette()` at the end. There's just one small wrinkle: you probably don't want to show the cassette usage in the example (because the user won't need it in their code), so you'll want to wrap the inserting and ejecting functions `\dontshow{}`.
+  resp_body_json(resp)
+}
 
-All together, this will lead to roxygen2 code that looks something like this:
+test_that("my test", {
+  vcr::local_vcr_configure_log(file = stdout())
+  vcr::local_cassette("body", match_requests_on = c("method", "uri", "body_json"))
+  
+  data <- get_data(b = 1)
+  expect_named(data$json, c("x", "y"))
+})
+```
 
 ```{r}
-#' @examples
-#' \dontshow{
-#' vcr::insert_example_cassette("casette-name", "package-name")
-#' }
-#' 
-#' output <- my_api_calling_function()
-#' str(output)
-#' 
-#' \dontshow{
-#' vcr::eject_cassette()
-#' }
+#| label: body-replay
+#| error: true
+
+test_that("my test", {
+  vcr::local_vcr_configure_log(file = stdout())
+  vcr::local_cassette("body", match_requests_on = c("method", "uri", "body_json"))
+  
+  data <- get_data(b = 2)
+  expect_named(data$json, c("x", "y"))
+})
 ```
 
-You'll then need to step through the example line by line to record the casssette for the first time.
+If you look at the log you'll note that we see exactly where the difference is in the json, even though it's buried several layers deep.
+Compare this to just using `"body"`, where you have to carefully compare two strings.
 
-Remember that every casssette used in an example has to be stored inside your package, making it larger, so we recommend that you use this technique sparingly.
+```{r}
+#| label: body-replay-raw
+#| error: true
+
+test_that("my test", {
+  vcr::local_vcr_configure_log(file = stdout())
+  vcr::local_cassette("body", match_requests_on = c("method", "uri", "body"))
+  
+  data <- get_data(b = 2)
+  expect_named(data$json, c("x", "y"))
+})
+```
```

Comment:
I know tutorials don't have to end with a conclusion, but this feels a bit abrupt?
If the sentence about using vcr in vignettes and examples were moved here, with a dedicated heading, would it be better?